### DISCUSSION TITLE
raylib v3.7.0

### DISCUSSION THREAD

**Discussion by raysan5** [raysan5/raylib#1736] [General]
April 2021, it's been about 4 months since last raylib release and here it is already a new one, this time with a bunch of internal redesigns and improvements. Surprisingly, on April the 8th I was awarded for a second time with the [Google Open Source Peer Bonus Award](https://opensource.googleblog.com/2021/04/announcing-first-group-of-google-open-source-peer-bonus-winners.html) for my contribution to open source world with raylib and it seems the library is getting some traction, what a better moment for a new release? Let's see what can be found in this new version:

Let's start with some numbers:

 - **+100** closed issues (for a TOTAL of **+900**!)
 - **+400** commits since previous RELEASE
 - **+50** functions ADDED (**+30** of them to rlgl API)
 - **+30** functions REVIEWED/REDESIGNED
 - **+40** new contributors (for a TOTAL of **+210**!)
 
Highlights for `raylib 3.7`:

 - **REDESIGNED: `rlgl` module for greater abstraction level**. This suppose an **important change in raylib architecture**, now `rlgl` functionality is self-contained in the module and used by higher-level layers (specially by `core` module), those upper layers are the ones that expose functionality to the main API when required, for example the `Shaders`, `Mesh` and `Materials` functionality. Multiple `rlgl` functions have been renamed for consistency, in this case, following the `rl*()` prefix convention. Functions have also been reorganized internally by categories and `GenTexture*()` functions have been removed from the library and moved to [`models_material_pbr`](https://github.com/raysan5/raylib/blob/master/examples/models/models_material_pbr.c) example.
 
 - **REDESIGNED: VR simulator and stereo rendering mechanism**. A **brand new API** has been added, more comprehensive and better integrated with raylib, the **new stereo rendering** can be combined with `RenderTexture` and `Shader` API allowing the user to **manage fbo and distortion shader directly**. Also, the new rendering mechanism supports **instancing on stereo rendering**! Check the updated [`core_vr_simulator`](https://github.com/raysan5/raylib/blob/master/examples/core/core_vr_simulator.c) example for reference!
 
 - **ADDED: New file access callbacks system**. Several new callback functions have been added to the API to allow custom file loaders. A [nice example](https://github.com/RobLoach/raylib-physfs) it's the **raylib integration with a virtual file system** [PhysFS](https://icculus.org/physfs/).
 
 - **ADDED: glTF animations support**. glTF is the preferred models file format to be used with raylib and along the addition of a models animation API on latest raylib versions, now animations support for glTF format has come to raylib, thanks for this great contribution to [Hristo Stamenov](@object71)
 
 - **ADDED: Music streaming support from memory**. raylib has been adding the `Load*FromMemory()` option to all its supported file formats but **music streaming** was not supported yet... until now. Thanks to this great contribution by [Agnis "NeZvērs" Aldiņš](@nezvers), now raylib supports music streamming from memory data for all supported file formats: WAV, OGG, MP3, FLAC, XM and MOD.
 
 - **RENAMED: enums values for consistency**. Most raylib enums names and values names have been renamed for consistency, now all value names start with the type of data they represent. It increases clarity and readability when using those values and also **improves overall library consistency**.
 
Beside those key changes, many functions have been reviewed with improvements and bug fixes, many of them contributed by the community! Thanks! And again, this release sets a **new milestone for raylib library**. Make sure to check [CHANGELOG](https://github.com/raysan5/raylib/blob/master/CHANGELOG) for detailed list of changes! Hope you enjoy this new raylib installment!

Happy **gamedev/tools/graphics** programming! :)

<hr /><em>This discussion was created from the release <a href='https://github.com/raysan5/raylib/releases/tag/3.7.0'>raylib v3.7.0</a>.</em>

**Comment by mkupiec1** [comment]
Hi,
Great job with whole library!
Just saw, that new function CheckCollisionPointLine is missing in Cheatsheet.
Bye!

**Comment by orcmid** [comment]
I did not know abou raylib, even though I was a big fan of BGI in the day and was always annoyed when Microsoft did not provide something comparable in Visual Studio.  

I learned of raylib because it is featured in a new Udemy course by Stephen Ulibarri: "[C++ Fundamentals:  Game Programming for Beginners](https://www.udemy.com/course/cpp-fundamentals/)."  

The course relies on Visual Studio Code on all platforms, with the `raylib_installer_v370.mingw` for windows.  There have been a few avalanches/train-wrecks in getting integrations into VSCode for Mac OS and Linux too.  Beginners are taken far out onto the skinny branches, so the course is being revised in real-time.

Meanwhile, I love the raylib approach and have been pouring over the code and looking into better integrations for operating with VS Code.  I can only do the Windows case (and maybe WSL for a Linux case) but I would love to get off of the MingW and g++, using the free Visual Studio Build Tools and a static library instead.  That will depend on the level of C99 support in VC.  

That's my itch and I intend to scratch it.  I also am delighted to see that Clean C is being used for C/C++ compiling.  

This might not be the best place to introduce myself, but I wanted to wave and also express my delight with raylib.

**Reply by raysan5** [reply]
Hello @orcmid! Glad you like raylib! :D I actually created it after using WinBGI for some weeks, I really liked it so I decided to create something similar using my experience with OpenGL and XNA library. I tested raylib with my students and they loved it so I kept improving it along the years. Actually, I started working on it on August 2013, 8 years ago!

Last time I checked the VSCode project was some time ago and maybe some things changed, I didn't know it was causing some trouble for the Udemy course... The main projects I maintain are the plain Makefiles and the VS2019 solution but I can put some effort to improve/review the VSCode one. Feel free to contribute!
